core/arch/arm/mm/mobj.c

```diff
@@ -312,6 +306,291 @@ struct mobj *mobj_mm_alloc(struct mobj *mobj_parent, size_t size,
+	mobj_reg_shm->cookie = cookie;
+	mobj_reg_shm->num_pages = num_pages;
+	mobj_reg_shm->page_offset = page_offset;
+	memcpy(mobj_reg_shm->pages, pages, sizeof(*pages) * num_pages);
+
+	exceptions = cpu_spin_lock_xsave(&reg_shm_slist_lock);
+	SLIST_INSERT_HEAD(&reg_shm_list, mobj_reg_shm, next);
+	cpu_spin_unlock_xrestore(&reg_shm_slist_lock, exceptions);
+
+	return &mobj_reg_shm->mobj;
+}
+
+struct mobj *mobj_reg_shm_find_by_cookie(uint64_t cookie)
+{
+	struct mobj_reg_shm *mobj_reg_shm;
+	unsigned int exceptions;
+
+	exceptions = cpu_spin_lock_xsave(&reg_shm_slist_lock);
+	SLIST_FOREACH(mobj_reg_shm, &reg_shm_list, next) {
+		if (mobj_reg_shm->cookie == cookie) {
+			cpu_spin_unlock_xrestore(&reg_shm_slist_lock,
+						 exceptions);
+			return &mobj_reg_shm->mobj;
+		}
+	}
+	cpu_spin_unlock_xrestore(&reg_shm_slist_lock, exceptions);
+	return NULL;
+}
+
+/*
+ * mobj_mapped_shm implementation.
+ * Used for registered SHM that is mapped to OP-TEE virtual address space
+ */
+
+struct mobj_mapped_shm {
+	struct mobj mobj;
+	tee_mm_entry_t *mm_entry;
+	struct mobj *reg_shm;
+	bool owns_reg_shm;
+};
+
+static struct mobj_mapped_shm *to_mobj_mapped_shm(struct mobj *mobj);
+
+static TEE_Result mobj_mapped_shm_get_pa(struct mobj *mobj, size_t offst,
+				      size_t granule, paddr_t *pa)
+{
+	struct mobj_mapped_shm *mobj_mapped_shm = to_mobj_mapped_shm(mobj);
+
+	return mobj_reg_shm_get_pa(mobj_mapped_shm->reg_shm, offst,
+				   granule, pa);
+}
+
+static void *mobj_mapped_shm_get_va(struct mobj *mobj, size_t offst)
+{
+	struct mobj_mapped_shm *mobj_mapped_shm = to_mobj_mapped_shm(mobj);
+	struct mobj_reg_shm *reg_shm =
+		to_mobj_reg_shm(mobj_mapped_shm->reg_shm);
+
+	return (void *)(tee_mm_get_smem(mobj_mapped_shm->mm_entry) + offst +
+			reg_shm->page_offset);
+}
+
+static void mobj_mapped_shm_free(struct mobj *mobj)
+{
+	struct mobj_mapped_shm *mobj_mapped_shm = to_mobj_mapped_shm(mobj);
+	vaddr_t va = tee_mm_get_smem(mobj_mapped_shm->mm_entry);
+
+	core_mmu_unmap_pages(va, mobj->size / SMALL_PAGE_SIZE);
+	tee_mm_free(mobj_mapped_shm->mm_entry);
+	if (mobj_mapped_shm->owns_reg_shm)
+		mobj_free(mobj_mapped_shm->reg_shm);
+	free(mobj_mapped_shm);
+}
+
+static TEE_Result mobj_mapped_shm_get_cattr(struct mobj *mobj __unused,
+					 uint32_t *cattr)
+{
+	if (!cattr)
+		return TEE_ERROR_GENERIC;
+
+	*cattr = TEE_MATTR_CACHE_CACHED;
+
+	return TEE_SUCCESS;
+}
+
+static bool mobj_mapped_shm_matches(struct mobj *mobj, enum buf_is_attr attr);
+
+static const struct mobj_ops mobj_mapped_shm_ops __rodata_unpaged = {
+	.get_pa = mobj_mapped_shm_get_pa,
+	.get_va = mobj_mapped_shm_get_va,
+	.get_cattr = mobj_mapped_shm_get_cattr,
+	.matches = mobj_mapped_shm_matches,
+	.free = mobj_mapped_shm_free,
+};
+
+static bool mobj_mapped_shm_matches(struct mobj *mobj __maybe_unused,
+				   enum buf_is_attr attr)
+{
+	assert(mobj->ops == &mobj_mapped_shm_ops);
+
+	return attr == CORE_MEM_NON_SEC || attr == CORE_MEM_EXTRAM ||
```

Comment:
`CORE_MEM_EXTRAM` cover the physical DDR. It may cover secure memory regions.
Can't we stick to `attr == CORE_MEM_NON_SEC || attr == CORE_MEM_NSEC_SHM`?